File: TypeScript/_9_linked_list/_21_merge_two_sorted_lists.js
var ListNode = /** @class */ (function () {
    function ListNode(val, next) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
    ListNode.prototype.toString = function () {
        var result = "[";
        var current = this;
        while (current != null) {
            result += current.val.toString();
            if (current.next != null) {
                result += " -> ";
            }
            current = current.next;
        }
        result += "]";
        return result;
    };
    return ListNode;
}());
function mergeTwoLists(list1, list2) {
    if (!list1)
        return list2;
    if (!list2)
        return list1;
    var dummy = new ListNode();
    var current = dummy;
    while (list1 && list2) {
        if (list1.val < list2.val) {
            current.next = list1;
            list1 = list1.next;
        }
        else {
            current.next = list2;
            list2 = list2.next;
        }
        current = current.next;
    }
    if (list1)
        current.next = list1;
    if (list2)
        current.next = list2;
    return dummy.next;
}
function testMergeTwoLists(list1, list2) {
    console.log("list1: ".concat(list1));
    console.log("list2: ".concat(list2));
    var start = new Date().getTime();
    var result = mergeTwoLists(list1, list2);
    var end = new Date().getTime();
    console.log("result: ".concat(result));
    console.log("Execution time: ".concat(end - start));
}
testMergeTwoLists(new ListNode(1, new ListNode(2, new ListNode(4))), new ListNode(1, new ListNode(3, new ListNode(4))));
testMergeTwoLists(null, null);
testMergeTwoLists(null, new ListNode(0));

File: Kotlin/src/_9_linked_list/_21_Merge_Two_Sorted_Lists.kt
package _9_linked_list

class ListNode(var `val`: Int) {

    var next: ListNode? = null

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("[")
        var current: ListNode? = this
        while (current != null) {
            sb.append(current.`val`)
            if (current.next != null)
                sb.append(" -> ")
            current = current.next
        }
        sb.append("]")
        return sb.toString()
    }
}

fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
    if (list1 == null) return list2
    if (list2 == null) return list1

    var head: ListNode? = null
    var current: ListNode? = null
    var l1 = list1
    var l2 = list2

    while (l1 != null && l2 != null) {
        if (l1.`val` < l2.`val`) {
            if (head == null) {
                head = l1
                current = head
            } else {
                current?.next = l1
                current = current?.next
            }
            l1 = l1.next
        } else {
            if (head == null) {
                head = l2
                current = head
            } else {
                current?.next = l2
                current = current?.next
            }
            l2 = l2.next
        }
    }

    if (l1 != null)
        current?.next = l1
    else
        current?.next = l2

    return head
}

fun testMergeTwoLists(list1: ListNode?, list2: ListNode?) {
    println("\nlist1: $list1")
    println("list2: $list2")

    val inicio = System.nanoTime()
    val result = mergeTwoLists(list1, list2)
    val fim = System.nanoTime()

    println("Time: ${fim - inicio}")
    println("result: $result")
}

fun createLinkedList(vararg values: Int): ListNode? {
    var head: ListNode? = null
    var tail: ListNode? = null
    for (value in values) {
        val newNode = ListNode(value)
        if (head == null) {
            head = newNode
        } else {
            tail?.next = newNode
        }
        tail = newNode
    }
    return head
}

fun main() {
    val list1 = createLinkedList(1, 2, 4)
    val list2 = createLinkedList(1, 3, 4)

    testMergeTwoLists(list1, list2)
}